### PR TITLE
[telemetry] add proto for DHCPv6 PD metrics

### PR DESCRIPTION
**Pull request by sherysheng**
Depends-On: https://github.com/openthread/openthread/pull/9645

### CODE REVIEW

File: src/proto/thread_telemetry.proto [hunk 1]
```diff
@@ -219,17 +219,45 @@ message TelemetryData {
...
+    optional Nat64ProtocolCounters nat64_protocol_counters = 21;
 
     // Error counters for NAT64 translator on the border router
-    optional Nat64ErrorCounters nat64_error_counters = 20;
+    optional Nat64ErrorCounters nat64_error_counters = 22;
+  }
+
+  enum Dhcp6PdState {
+    // DHCPv6 PD is disabled on the border router.
+    DHCP6_PD_STATE_DISABLED = 0;
```
Comment (erjiaqing) [inline src/proto/thread_telemetry.proto:243]: Add DHCP6_PD_STATE_UNSPECIFIED for missing / default value.

File: src/utils/thread_helper.cpp [hunk 2]
```diff
@@ -184,6 +184,23 @@ void CopyNat64TrafficCounters(const otNat64Counters &from, threadnetwork::Teleme
...
+{
+    switch (dhcp6PdState)
+    {
+    case OT_BORDER_ROUTING_DHCP6_PD_STATE_DISABLED:
+        return threadnetwork::TelemetryData::DHCP6_PD_STATE_DISABLED;
+    case OT_BORDER_ROUTING_DHCP6_PD_STATE_STOPPED:
+        return threadnetwork::TelemetryData::DHCP6_PD_STATE_STOPPED;
+    case OT_BORDER_ROUTING_DHCP6_PD_STATE_RUNNING:
+        return threadnetwork::TelemetryData::DHCP6_PD_STATE_RUNNING;
+    default:
```
Comment (erjiaqing) [inline src/utils/thread_helper.cpp:198]: Remove the default, so the compiler will block the future changes.

**Review by erjiaqing — APPROVED** [review]

File: src/utils/thread_helper.cpp [hunk 3]
```diff
@@ -184,6 +184,23 @@ void CopyNat64TrafficCounters(const otNat64Counters &from, threadnetwork::Teleme
...
+        return threadnetwork::TelemetryData::DHCP6_PD_STATE_UNSPECIFIED;
```
Comment (Irving-cl) [inline src/utils/thread_helper.cpp:199]:
```suggestion
    TelemetryData_Dhcp6PdState pdState = TelemetryData::DHCP6_PD_STATE_UNSPECIFIED;

    switch (dhcp6PdState)
    {
    case OT_BORDER_ROUTING_DHCP6_PD_STATE_DISABLED:
        pdState = threadnetwork::TelemetryData::DHCP6_PD_STATE_DISABLED;
    case OT_BORDER_ROUTING_DHCP6_PD_STATE_STOPPED:
        pdState = threadnetwork::TelemetryData::DHCP6_PD_STATE_STOPPED;
    case OT_BORDER_ROUTING_DHCP6_PD_STATE_RUNNING:
        pdState = threadnetwork::TelemetryData::DHCP6_PD_STATE_RUNNING;
    default:
        break;
    }
    
    return pdState;
```

Change to the "one return statement" style? I understand there are some code in this file that don't follow this style. But I think OpenThread follows this style. So we'd better also keep it in ot-br-posix.

File: third_party/openthread/repo
[the remark this answers is not in the document]
Comment (erjiaqing) [reply third_party/openthread/repo:1]: The bump is used for building and testing, the change will be reverted once the openthread PR is merged.

**Comment by sherysheng** [comment]
> Depends-On: [openthread/openthread#9645](https://github.com/openthread/openthread/pull/9645)

The Dependency has been merged, so the sumodule commit has been removed now

**Review by Irving-cl — APPROVED** [review]
LGTM 👍

File: src/utils/thread_helper.cpp [hunk 4]
```diff
@@ -202,10 +226,10 @@ ThreadHelper::ThreadHelper(otInstance *aInstance, otbr::Ncp::ControllerOpenThrea
     : mInstance(aInstance)
     , mNcp(aNcp)
 {
-#if OTBR_ENABLE_TELEMETRY_DATA_API && OTBR_ENABLE_NAT64
+#if OTBR_ENABLE_TELEMETRY_DATA_API && OTBR_ENABLE_NAT64 || OTBR_ENABLE_DHCP6_PD
```
Comment (morningboata) [inline src/utils/thread_helper.cpp:229]: Do you want to wrap the condition like (OTBR_ENABLE_NAT64 || OTBR_ENABLE_DHCP6_PD)?

File: src/utils/thread_helper.cpp [hunk 5]
```diff
@@ -1360,8 +1392,48 @@ otError ThreadHelper::RetrieveTelemetryData(Mdns::Publisher *aPublisher, threadn
...
+
+            sha256.Start();
+            sha256.Update(pdPrefixShaInput, sizeof(pdPrefixShaInput));
+            sha256.Finish(hash);
+
+            wpanBorderRouter->mutable_hashed_pd_prefix()->append(reinterpret_cast<const char *>(hash.GetBytes()),
+                                                                 sizeof(hash.GetBytes()));
+        }
+        // End of PD prefix
+        // Start of DHCPv6 PD processed RA Info
```
Comment (morningboata) [inline src/utils/thread_helper.cpp:1424]: // Start of DHCPv6 PdProcessedRaInfo section.
Use same name as in proto may help code search slightly easier.

Comment (sherysheng) [reply src/utils/thread_helper.cpp:1424]: Thanks!

File: src/utils/thread_helper.cpp [hunk 6]
```diff
@@ -1360,8 +1392,48 @@ otError ThreadHelper::RetrieveTelemetryData(Mdns::Publisher *aPublisher, threadn
...
+                   sizeof(aPrefixInfo.mPrefix.mPrefix.mFields.m8));
+            memcpy(&pdPrefixShaInput[sizeof(aPrefixInfo.mPrefix.mPrefix.mFields.m8)], mNat64PdCommonSalt,
+                   sizeof(mNat64PdCommonSalt));
+
+            sha256.Start();
+            sha256.Update(pdPrefixShaInput, sizeof(pdPrefixShaInput));
+            sha256.Finish(hash);
+
+            wpanBorderRouter->mutable_hashed_pd_prefix()->append(reinterpret_cast<const char *>(hash.GetBytes()),
+                                                                 sizeof(hash.GetBytes()));
```
[the remark this answers is not in the document]
Comment (sherysheng) [reply src/utils/thread_helper.cpp:1421]: I'm using the bytes size here, isn't it the same?

[the remark this answers is not in the document]
Comment (superwhd) [reply src/utils/thread_helper.cpp:1421]: It seems the issue also existed for the NAT64 telemetry. @morningboata

[the remark this answers is not in the document]
Comment (abtink) [reply src/utils/thread_helper.cpp:1421]: @superwhd  is correct. The use of `sizeof()` does not do what you want (which is the number of bytes in the `hash` array).

`sizeof()` returns the size of an object or a type.  In the case here, `hash.GetBytes()` is a method call, so compiler replaces it with the expected return type of `GetBytes()` method which is defined as 
```c++
const uint8_t *GetBytes(void) const;
```` 

so the `sizeof(hash.GetBytes())` is the same as `sizeof(const uint8_t *)` which gives the size of a pointer variable and not the length of the array the pointer may be pointing to.

File: src/utils/thread_helper.cpp [hunk 7]
```diff
@@ -1360,8 +1392,48 @@ otError ThreadHelper::RetrieveTelemetryData(Mdns::Publisher *aPublisher, threadn
...
-        // End of WpanBorderRouter section.
+        // Start of Hashed PD prefix
+        {
+            otBorderRoutingPrefixTableEntry aPrefixInfo;
+            Sha256::Hash                    hash;
+            Sha256                          sha256;
+
+            otBorderRoutingGetPdOmrPrefix(mInstance, &aPrefixInfo);
+            uint8_t pdPrefixShaInput[OT_IP6_ADDRESS_SIZE + kNat64PdCommonHashSaltLength];
+            memcpy(pdPrefixShaInput, aPrefixInfo.mPrefix.mPrefix.mFields.m8,
```
Comment (abtink) [inline src/utils/thread_helper.cpp:1411]: Suggestion to simplify this code:
- We dont need to `memcpy` to construct the whole `pdPrefixShaInput` buffer by appending the prefix first followed by the slat.
- The `Sha256::Update()` method can be called multiple times.
- We can read the prefix, feed it directly in `Sha256::Update()`
- Then call `Sha256::Update()` again to feed the salt bytes.

File: src/utils/thread_helper.cpp [hunk 8]
```diff
@@ -1360,8 +1392,48 @@ otError ThreadHelper::RetrieveTelemetryData(Mdns::Publisher *aPublisher, threadn
...
+                   sizeof(aPrefixInfo.mPrefix.mPrefix.mFields.m8));
```
Comment (abtink) [inline src/utils/thread_helper.cpp:1412]: PD prefix will be shorter than 128 bits long, but here we are using the entire `OT_IP6_ADDRESS_SIZE`. Is this correct/desired?

Comment (sherysheng) [reply src/utils/thread_helper.cpp:1412]: Updated according to [go/thread-ipaddress-obfuscation](https://goto.google.com/thread-ipaddress-obfuscation)

File: src/utils/thread_helper.cpp [hunk 9]
```diff
@@ -1343,25 +1379,85 @@ otError ThreadHelper::RetrieveTelemetryData(Mdns::Publisher *aPublisher, threadn
...
+#if OTBR_ENABLE_DHCP6_PD
+        // Start of Dhcp6PdState section.
+        {
+            wpanBorderRouter->set_dhcp6_pd_state(
+                Dhcp6PdStateFromOtDhcp6PdState(otBorderRoutingDhcp6PdGetState(mInstance)));
+        }
+        // End of Dhcp6PdState section.
+
+        // Start of Hashed PD prefix
+        {
```
Comment (superwhd) [inline src/utils/thread_helper.cpp:1408]: In the following block, please group all the variable declarations and put them at the beginning of the block.

File: src/utils/thread_helper.cpp [hunk 10]
```diff
@@ -1343,25 +1379,85 @@ otError ThreadHelper::RetrieveTelemetryData(Mdns::Publisher *aPublisher, threadn
...
+            otBorderRoutingPrefixTableEntry aPrefixInfo;
+
+            otBorderRoutingGetPdOmrPrefix(mInstance, &aPrefixInfo);
+
+            const uint8_t       *prefixAddr          = aPrefixInfo.mPrefix.mPrefix.mFields.m8;
+            const size_t         kHashPrefixLength   = 6;
+            const size_t         kHashedPrefixLength = 2;
+            std::vector<uint8_t> hashed_pd_header    = {0x20, 0x01, 0x0d, 0xb8};
```
Comment (superwhd) [inline src/utils/thread_helper.cpp:1416]: nit: constexpr

File: src/utils/thread_helper.hpp [hunk 11]
```diff
@@ -332,9 +332,9 @@ class ThreadHelper
     UpdateMeshCopTxtHandler mUpdateMeshCopTxtHandler;
 #endif
 
-#if OTBR_ENABLE_TELEMETRY_DATA_API & OTBR_ENABLE_NAT64
-    static const uint8_t kNat64SourceAddressHashSaltLength = 16;
-    uint8_t              mNat64Ipv6AddressSalt[kNat64SourceAddressHashSaltLength];
+#if OTBR_ENABLE_TELEMETRY_DATA_API && (OTBR_ENABLE_NAT64 || OTBR_ENABLE_DHCP6_PD)
+    static const uint8_t kNat64PdCommonHashSaltLength = 16;
```
Comment (superwhd) [inline src/utils/thread_helper.hpp:336]: nit: constexpr

File: src/utils/thread_helper.cpp [hunk 12]
```diff
@@ -1343,25 +1379,85 @@ otError ThreadHelper::RetrieveTelemetryData(Mdns::Publisher *aPublisher, threadn
...
+            std::vector<uint8_t> hashed_pd_tailer    = {0x00, 0x00, 0x00, 0x00, 0x00, 0x00};
+            std::vector<uint8_t> hashed_pd_prefix;
```
Comment (superwhd) [inline src/utils/thread_helper.cpp:1418]: please use camel naming style

File: src/utils/thread_helper.cpp [hunk 13]
```diff
@@ -1343,25 +1377,86 @@ otError ThreadHelper::RetrieveTelemetryData(Mdns::Publisher *aPublisher, threadn
...
+
+            // Append ip[6] and ip[7]
+            hashedPdPrefix.push_back(prefixAddr[6]);
+            hashedPdPrefix.push_back(prefixAddr[7]);
+
+            // Append hashedPdTailer
+            hashedPdPrefix.insert(hashedPdPrefix.end(), hashedPdTailer.begin(), hashedPdTailer.end());
+
+            // Convert the vector to a string
+            std::string result(hashedPdPrefix.begin(), hashedPdPrefix.end());
```
Comment (superwhd) [inline src/utils/thread_helper.cpp:1442]: Variable declaration should be grouped together at the beginning of the code block.

BTW, this variable is not so necessary?  Probably you can do this:
```
wpanBorderRouter->set_hashed_pd_prefix(std::string(hashedPdPrefix.begin(), hashedPdPrefix.end()));
```

**Review by superwhd — APPROVED** [review]
LGTM with a nit.

File: src/utils/thread_helper.cpp (hunk 13 above)
Comment (sherysheng) [reply src/utils/thread_helper.cpp]: Thanks!

File: src/proto/thread_telemetry.proto [hunk 14]
```diff
@@ -431,6 +461,15 @@ message TelemetryData {
 
     // Information about the mappings of NAT64 translator
     repeated Nat64Mapping nat64_mappings = 7;
+
+    // DHCPv6 PD state
+    optional Dhcp6PdState dhcp6_pd_state = 8;
+
+    // DHCPv6 PD prefix
+    optional string hashed_pd_prefix = 9;
```
[the remark this answers is not in the document]
Comment (sherysheng) [reply src/proto/thread_telemetry.proto:469]: I think string is more feasible here, any strong reason we need to use bytes?

File: tests/dbus/test_dbus_client.cpp [hunk 15]
```diff
@@ -285,6 +285,9 @@ void CheckTelemetryData(ThreadApiDBus *aApi)
 #if OTBR_ENABLE_NAT64
     TEST_ASSERT(telemetryData.wpan_border_router().nat64_state().prefix_manager_state() ==
                 threadnetwork::TelemetryData::NAT64_STATE_NOT_RUNNING);
+#endif
+#if OTBR_ENABLE_DHCP6_PD
```
[the remark this answers is not in the document]
Comment (sherysheng) [reply tests/dbus/test_dbus_client.cpp:289]: How can I confirm this?

File: src/utils/thread_helper.cpp [hunk 16]
```diff
@@ -1343,25 +1377,84 @@ otError ThreadHelper::RetrieveTelemetryData(Mdns::Publisher *aPublisher, threadn
...
+        {
+            wpanBorderRouter->set_dhcp6_pd_state(
+                Dhcp6PdStateFromOtDhcp6PdState(otBorderRoutingDhcp6PdGetState(mInstance)));
+        }
+        // End of Dhcp6PdState section.
 
-        // End of WpanBorderRouter section.
+        // Start of Hashed PD prefix
+        {
+            otBorderRoutingPrefixTableEntry aPrefixInfo;
```
[the remark this answers is not in the document]
Comment (sherysheng) [reply src/utils/thread_helper.cpp:1407]: Will do

File: tests/dbus/test_dbus_client.cpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (sherysheng) [reply tests/dbus/test_dbus_client.cpp:289]: https://github.com/openthread/ot-br-posix/actions/runs/7138804450/job/19440989550?pr=2110  
It's covered in script-check

File: script/test [hunk 17]
```diff
@@ -130,6 +130,7 @@ do_build()
         "-DCMAKE_INSTALL_PREFIX=/usr"
         "-DOT_THREAD_VERSION=1.3"
         "-DOTBR_DBUS=ON"
+        "-DOTBR_DHCP6_PD=ON"
```
Comment (erjiaqing) [inline script/test:133]: You should also enable OTBR_BORDER_ROUTING if you enable OTBR_DHCP6_PD

**Review by abtink — APPROVED** [review]
LGTM. Couple of smaller suggestions below.

**Review by morningboata — APPROVED** [review]

File: src/proto/thread_telemetry.proto [hunk 18]
```diff
@@ -230,6 +230,36 @@ message TelemetryData {
...
+    optional PacketsAndBytes outbound_internet = 22;
+  }
+
+  enum Dhcp6PdState {
+    DHCP6_PD_STATE_UNSPECIFIED = 0;
+
+    // DHCPv6 PD is disabled on the border router.
+    DHCP6_PD_STATE_DISABLED = 1;
+
+    // DHCPv6 PD in enabled but won't try to request and publish a prefix.
```
Comment (jwhui) [inline src/proto/thread_telemetry.proto:247]:
```suggestion
    // DHCPv6 PD is enabled but won't try to request and publish a prefix.
```

File: src/utils/thread_helper.cpp [hunk 19]
```diff
@@ -1342,26 +1375,77 @@ otError ThreadHelper::RetrieveTelemetryData(Mdns::Publisher *aPublisher, threadn
...
+            const uint8_t                  *prefixAddr          = nullptr;
+            const uint8_t                  *truncatedHash       = nullptr;
+            constexpr size_t                kHashPrefixLength   = 6;
+            constexpr size_t                kHashedPrefixLength = 2;
+            std::vector<uint8_t>            hashedPpdHeader     = {0x20, 0x01, 0x0d, 0xb8};
```
Comment (jwhui) [inline src/utils/thread_helper.cpp:1402]: What does `Ppd` stand for?

Comment (sherysheng) [reply src/utils/thread_helper.cpp:1402]: typo, updated! Thanks

**Review by jwhui — APPROVED** [review]

**Pull request merged**